### PR TITLE
Implementing Offline Caching In Room For Data Fetched From Firebase Using Kotlin

### LINKED ISSUE #5107: [Android] Implementing Offline Caching In Room For Data Fetched From Firebase Using Kotlin
### Topic Suggestion
[Android] Implementing Offline Caching With Room For Data Fetched From Firebase Using Kotlin

#### Proposed title of article
Implementing Offline Caching With Room For Data Fetched From Firebase Using Kotlin

#### Proposed article introduction
The Room persistence library provides an abstraction layer over SQLite to allow fluent database access while harnessing the full power of SQLite.

The most common use case is to cache relevant pieces of data so that when the device cannot access the network, the user can still browse that content while they are offline.

#### Key takeaways 
- How to use Room database
- How to cache data offline using Room.

#### Article quality
- It is user friendly for beginners  since it is easy to read and understand. The contents are arranged and organized in a way such that the reader will be able to gather crucial information on caching with room with ease. 
- This article will explain how to cache data fetched from database using Room.

#### References
- [Offline Caching In Room](https://developer.android.com/training/data-storage/room)

### PR DESCRIPTION
**Pull request by Sephens**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5107

### CODE REVIEW

**Comment by lalith1403** [comment]
plagiarism check done ✔️

File: content/authors/stephen-odhiambo/index.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+---
+title: Stephen Odhiambo
+type: authors
+github: https://github.com/Sephens
+---
```
Comment (Dawe-7) [inline content/authors/stephen-odhiambo/index.md:5]:
```suggestion
title: Stephen Odhiambo
type: authors
github: https://github.com/Sephens
images:
  - url: /engineering-education/authors/stephen-odhiambo/avatar.jpeg
skills: ["Kotlin" etc.]
---

File: content/articles/firebase-offline-caching/index.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+    val itemName: String? = null,
+    val itemOldPrice: String? = null,
+    val itemPrice: String? = null
+)
+```
+
+#### Step 2 - Defining utility items
+In the `util` package, define the following utility items. The utility package is for defining utility methods constants that can be reused in various classes.
+
+```kotlin
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:118]: Add the missing code here

File: content/articles/firebase-offline-caching/index.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+```
+This sealed class will help us in handling the three states of our real-time database network calls i.e `Loading`, `Success`, and `Error`.
+
+#### Step 3 - Creating data classes
+We will define our `DAO` and `Database` classes in the `data` package.
+
+#### Creating DAO interface
+DAO is an interface that has methods for manipulating the local database. It should look like the one below:
+
+@Database(
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:129]: Add backticks and use the proper language/code identifier. Ensure all your code runs perfectly.

File: content/articles/firebase-offline-caching/index.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+    fun getAllItems(): Flow<List<ItemsEntity>>
+    
+    // Method for searching an item from the DB
+    @Query("SELECT * FROM items WHERE itemName LIKE :searchQuery")
+    fun searchDatabase(searchQuery: String): Flow<List<ItemsEntity>>
+}
+```
+
+#### Creating database class
+This class has a single method that returns the object of our DAO.
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:154]:
```suggestion
This class has a single method that returns the object of our DAO. We create the database class using the code below:

File: content/articles/firebase-offline-caching/index.md [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+)
+abstract class ItemsDatabase: RoomDatabase() {
+    abstract fun getItemsDao(): ItemsDao
+}
+```
+
+#### Step 4 - Repository class
+This class will have the business logic for fetching the data and caching it for offline access. In the `repository` package, create a repository class and name it `ItemsRepository`.
+
+```kotlin
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:169]: Add the missing code here

File: content/articles/firebase-offline-caching/index.md [hunk 6]
```diff
@@ -0,0 +1,294 @@
...
+    entities = [ItemsEntity::class],
+    version = 1
+)
+abstract class ItemsDatabase: RoomDatabase() {
+    abstract fun getItemsDao(): ItemsDao
+}
+```
+
+#### Step 4 - Repository class
+This class will have the business logic for fetching the data and caching it for offline access. In the `repository` package, create a repository class and name it `ItemsRepository`.
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:167]:
```suggestion
This class will have the business logic for fetching the data and caching it for offline access. In the `repository` package, create a repository class and name it `ItemsRepository`. Use the code below to achieve this:
```

File: content/articles/firebase-offline-caching/index.md [hunk 7]
```diff
@@ -0,0 +1,294 @@
...
+                is Resource.Error<*> -> {
+                    binding.progressBar2.isVisible = false
+                    Snackbar.make(binding.root, result.message.toString(), Snackbar.LENGTH_LONG).show()
+                }
+            }
+        })
+    }
+}
+```
+![demo](/engineering-education/firebase-offline-caching/demo.jpeg)
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:286]:
```suggestion
![Demo](/engineering-education/firebase-offline-caching/demo.jpeg)
```

File: content/articles/firebase-offline-caching/index.md [hunk 8]
```diff
@@ -0,0 +1,294 @@
...
+#### Step 2 - Defining utility items
+In the `util` package, define the following utility items. The utility package is for defining utility methods constants that can be reused in various classes.
+
+```kotlin
+
+```
+This sealed class will help us in handling the three states of our real-time database network calls i.e `Loading`, `Success`, and `Error`.
+
+#### Step 3 - Creating data classes
+We will define our `DAO` and `Database` classes in the `data` package.
```
Comment (Dawe-7) [inline content/articles/firebase-offline-caching/index.md:124]: Add more content here.

**Review by Dawe-7 — CHANGES_REQUESTED** [review]
Hello @Sephens  , welcome to Section's EngEd and congratulations on submitting your first article. The article is a nice read but i noticed you haven't completely finished and a few things need to be corrected.

Make the following changes/adjustments:
- Ensure your hero image is of the size 600x400 pixels and is not blurred. Some of your images are too large, check on them and resize them.
- Check your article for readability and ensure you code runs. Kindly add the missing code in the article because i've seen a few missing ones.
- If any of your images are cited, kindly add image sources for them.

Feel free to ping me in case you get stuck, i'll be happy to help!

Kindly go through our [docs](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/UPLOAD_INSTRUCTIONS.md#create-author-bio-folder-structure) to get familiar with the process. Revisit your article once more and check on your grammar to make the review process smooth.

**Comment by hectorkambow** [comment]
closing this PR to take out of the queue while edits are being made

**Pull request closed without merging**